model/riscv_insts_zcmp.sail

```diff
@@ -0,0 +1,643 @@
+/*=======================================================================================*/
+/*  This Sail RISC-V architecture model, comprising all files and                        */
+/*  directories except where otherwise noted is subject the BSD                          */
+/*  two-clause license in the LICENSE file.                                              */
+/*                                                                                       */
+/*  SPDX-License-Identifier: BSD-2-Clause                                                */
+/*=======================================================================================*/
+
+
+enum clause extension = Ext_Zcmp
+function clause extensionEnabled(Ext_Zcmp) = extensionEnabled(Ext_Zca) & not(extensionEnabled(Ext_Zcd))
+
+mapping RLIST_MAPPING : bits(4) <-> string = {
+  0x4 <-> "{ra}",
+  0x5 <-> "{ra, s0}",
+  0x6 <-> "{ra, s0-s1}",
+  0x7 <-> "{ra, s0-s2}",
+  0x8 <-> "{ra, s0-s3}",
+  0x9 <-> "{ra, s0-s4}",
+  0xa <-> "{ra, s0-s5}",
+  0xb <-> "{ra, s0-s6}",
+  0xc <-> "{ra, s0-s7}",
+  0xd <-> "{ra, s0-s8}",
+  0xe <-> "{ra, s0-s9}",
+  0xf <-> "{ra, s0-11}",
+} 
+
+mapping CM_PUSH_MAPPING_32 : (bits(4), bits(2)) <-> string = {
+  (0x4, 0b00) <-> "{ra}, -16",
```

Comment:
Hm, I feel like there must be a better way to do this reusing the rlist mapping, duplicating it all seems rather sad. Ideally the stack adjustment would also be calculated in a general manner rather than every case enumerated individually.

Reply:
Yes. Exactly. This can be done relatively straightforwardly. I think something like might work

```
type Zcmp_constants = {1, 4, 7}

enum zcmp_variants with as_nat -> Zcmp_constants = {
   RV32E_Zcmp => 1,
   RV32I_Zcmp => 4,
   RV64_Zcmp  => 7
}

type RList = bits(4)
type Spimm = bits(2)

function stack_adj_base(variant : zcmp_variants, rlist : RList) -> nat = {
  match (variant, unsigned(rlist)) {
    (RV32E_Zcmp, _) => 16,
    (RV32I_Zcmp, i) if 4 <= i  & i < 8  => 16,
    (RV32I_Zcmp, i) if 8 <= i  & i < 12 => 32,
    (RV32I_Zcmp, i) if 12 <= i & i < 15 => 48,
    (RV32I_Zcmp, 15)                    => 64,
    (RV64_Zcmp, i)  if 4 <= i  & i < 6  => 16,
    (RV64_Zcmp, i)  if 6 <= i  & i < 8  => 32,
    (RV64_Zcmp, i)  if 8 <= i  & i < 10 => 48,
    (RV64_Zcmp, i)  if 10 <= i & i < 12 => 64,
    (RV64_Zcmp, i)  if 12 <= i & i < 14 => 80,
    (RV64_Zcmp, 14)                     => 96,
    (RV64_Zcmp, 15)                     => 112,
    (_, _) => internal_error(__FILE__, __LINE__, "Using reserved bits ")

function stack_adj(variant : zcmp_variants, rlist : RList, spimm : Spimm) -> nat = {
  stack_adj_base(variant, rlist) + unsigned(spimm) * 16
}
```